shards/modules/physics/physics.cpp

```diff
@@ -0,0 +1,661 @@
+#include <shards/common_types.hpp>
+#include <shards/linalg_shim.hpp>
+#include <shards/gfx/types.hpp>
+#include <gfx/hasherxxh3.hpp>
+#include <gfx/drawable.hpp>
+#include <tracy/Wrapper.hpp>
+#include "physics.hpp"
+#include "core.hpp"
+
+#include <Jolt/RegisterTypes.h>
+#include <Jolt/Core/Factory.h>
+
+namespace shards::Physics {
+std::atomic_uint64_t BodyNode::UidCounter = 0;
+std::atomic_uint64_t SHShape::UidCounter;
+std::atomic_uint64_t SHSoftBodyShape::UidCounter;
+
+struct ContextShard {
+  static SHTypesInfo inputTypes() { return shards::CoreInfo::AnyType; }
+  static SHTypesInfo outputTypes() { return ShardsContext::Type; }
+  static SHOptionalString help() {
+    return SHCCSTR("The core of the physics system, needs to be activated at the start of the simulation");
+  }
+
+  PARAM_IMPL();
+
+  ShardsContext *_context{};
+
+  void warmup(SHContext *context) {
+    PARAM_WARMUP(context);
+    shassert(!_context);
+    _context = ShardsContext::ObjectVar.New();
+    _context->core = std::make_shared<Core>();
+    _context->core->init();
+  }
+  void cleanup(SHContext *context) {
+    PARAM_CLEANUP(context);
+    if (_context) {
+      ShardsContext::ObjectVar.Release(_context);
+      _context = nullptr;
+    }
+  }
+
+  PARAM_REQUIRED_VARIABLES();
+  SHTypeInfo compose(SHInstanceData &data) {
+    PARAM_COMPOSE_REQUIRED_VARIABLES(data);
+    return outputTypes().elements[0];
+  }
+
+  SHVar activate(SHContext *shContext, const SHVar &input) {
+    _context->core->begin();
+    return ShardsContext::ObjectVar.Get(_context);
+  }
+};
+
+struct EndContextShard {
+  static SHTypesInfo inputTypes() { return shards::CoreInfo::AnyType; }
+  static SHTypesInfo outputTypes() { return shards::CoreInfo::AnyType; }
+  static SHOptionalString help() { return SHCCSTR("Runs physics simulation, should be run after defining physics bodies"); }
+
+  PARAM_PARAMVAR(_context, "Context", "The context", {ShardsContext::VarType});
+  PARAM_PARAMVAR(_timeStep, "TimeStep", "Time to simulate", {CoreInfo::FloatType, CoreInfo::FloatVarType});
+  PARAM_PARAMVAR(_maxIterations, "MaxIterations", "Maximum number of iterations to run the simulation in",
+                 {CoreInfo::IntType, CoreInfo::IntVarSeqType});
+  PARAM_IMPL(PARAM_IMPL_FOR(_context), PARAM_IMPL_FOR(_timeStep), PARAM_IMPL_FOR(_maxIterations));
+
+  EndContextShard() {
+    _timeStep = Var(1.0f / 60.0f);
+    _maxIterations = Var(1);
+  }
+
+  void warmup(SHContext *context) { PARAM_WARMUP(context); }
+  void cleanup(SHContext *context) { PARAM_CLEANUP(context); }
+
+  PARAM_REQUIRED_VARIABLES();
+  SHTypeInfo compose(SHInstanceData &data) {
+    PARAM_COMPOSE_REQUIRED_VARIABLES(data);
+    return outputTypes().elements[0];
+  }
+
+  SHVar activate(SHContext *shContext, const SHVar &input) {
+    double timeStep = _timeStep.get().payload.floatValue;
+    auto numIterations = _maxIterations.get().payload.intValue;
+
+    if (numIterations < 1 || numIterations > 128)
```

Comment:
Use of magic numbers: Replace magic numbers (like 128 in EndContextShard) with named constants for better readability and maintainability.

Maybe?

Reply:
Did that